File: 程序员面试金典/4.5合法二叉搜索树/isValidBST_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 最小最大值法
class Solution {
    public boolean isValidBST(TreeNode root) {
        return checkBST(root, null, null);
    }

    private boolean checkBST(TreeNode root, Integer min, Integer max) {
        if (root == null)
            return true;
        if (min != null && root.val <= min || max != null && root.val >= max)
            return false;
        if (!checkBST(root.left, min, root.val) || !checkBST(root.right, root.val, max))
            return false;
        return true;
    }

}

File: Algorithms/98/isValidBST_2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 递归
class Solution:
    def isValidBST(self, root: Optional[TreeNode], lower=float('-inf'), upper=float('inf')) -> bool:
        return not root or lower < root.val < upper and self.isValidBST(root.left, lower, root.val) and self.isValidBST(root.right, root.val, upper)
